catalog/transformer/catalog-transformer-xml/src/main/java/ddf/catalog/transformer/xml/adapter/AdaptedMetacard.java

```diff
@@ -188,11 +189,11 @@ public final Attribute getAttribute(String name) {
         if (Metacard.ID.equals(name)) {
             return this.id;
         }
-        for (Attribute attribute : attributes) {
-            if (attribute == null || StringUtils.isEmpty(attribute.getName())) {
-                continue;
-            } else if (name.equals(attribute.getName())) {
-                return attribute;
+        for (Attribute attribute : Collections.unmodifiableList(attributes)) {
```

Comment:
One more time. Sorry Jason.

After a deeper dive, there is lingering concern that the defensive copy above itself could still produce a `ConcurrentModificationException`. That appears not to be a problem as of Java 6, but I'm still leery of it. A guaranteed defensive copy can be gotten by changing the line to

```java
for (Attribute attribute: Arrays.asList(attributes.toArray())) {
```

However, there are issues with the defensive copy regardless of which way it is achieved. Most notably, if the collection *is* changed from beneath this thread, the attributes being iterated will not represent the actual state of the metacard. A comment should be added indicating this small but non-zero concern. This issue, notably, would still exist with a Concurrent collection; the use of that would only protect against CMEs.

The only way, short of an immutable implementation, of resolving this completely would be to synchronize all access to the attribute list and impacting performance; however, as it leaks out for jaxb marshalling, that isn't really feasible.